### PR TITLE
External allocator and configurable bitstream buffer

### PR DESCRIPTION
**Pull request by kevincoyle2018**
Update SDecodingParam to accept a pointer to an external memory allocator.

Also allow the bitstream buffer to be configurable instead of being fixed to 3 MB.

### CODE REVIEW

File: codec/common/inc/memory_align.h [hunk 1]
```diff
@@ -34,9 +34,10 @@
 #define WELS_COMMON_MEMORY_ALIGN_H__
 
 #include "typedefs.h"
+#include "codec_app_def.h"
 
 // NOTE: please do not clean below lines even comment, turn on for potential memory leak verify and memory usage monitor etc.
-//#define MEMORY_CHECK
+#define MEMORY_CHECK
```
Comment (thaytan) [inline codec/common/inc/memory_align.h]: I don't think you meant to leave this on here.

File: codec/console/dec/src/h264dec.cpp [hunk 2]
```diff
@@ -479,8 +479,6 @@ int32_t main (int32_t iArgC, char* pArgV[]) {
   }
 
 
-
-
```
Comment (thaytan) [inline codec/console/dec/src/h264dec.cpp]: Unnecessary whitespace changes here in h264dec.cpp

File: codec/decoder/plus/src/welsDecoderExt.cpp [hunk 3]
```diff
@@ -234,19 +234,19 @@ void CWelsDecoder::UninitDecoder (void) {
...
+    pSMemAllocator->WelsFree (m_pDecContext, "m_pDecContext");
     m_pDecContext = NULL;
   }
+
+  if (NULL != pSMemAllocator) {
+    WelsLog (&m_pWelsTrace->m_sLogCtx, WELS_LOG_INFO,
+             "CWelsDecoder::UninitDecoder(), verify memory usage (%d bytes) after free..",
+             pSMemAllocator->WelsGetMemoryUsage());
+    delete pSMemAllocator;
+  }
```
Comment (thaytan) [inline codec/decoder/plus/src/welsDecoderExt.cpp]: You should set m_pDecContext->pMemAlign back to NULL here, as UninitDecoder() can be called multiple times potentially (if InitDecoder fails for some reason). You therefore also can't assume that pSMemAllocator is non-NULL on entry.

**Comment by kevincoyle2018** [comment]
External allocation allows access to heaps not part of system memory.  For example, some silicon vendors may partition physical memory where X is available via standard Linux calls and reserve the rest for things like hardware video decoder buffers and so on.  However, since this partitioning occurs at device deployment time (think a set top box or TV) and is not adjustable dynamically, this memory may go to waste.  These vendors would have an API to allocate/deallocate/map that memory.

Additionally, it allows introspection into memory usage instead of relying on a developer to use a specialized tool (build of h264dec with MEMORY_CHECK).  For example, the allocator may be queriable via some JS bridge allowing a JS developer to determine quickly that memory usage is  higher than some planned budget and empower him to speak with an in-house content creator.

File: codec/console/dec/src/h264dec.cpp (hunk 2 above)
Comment (kevincoyle2018) [reply codec/console/dec/src/h264dec.cpp]: Thanks, restored to original state.
